---
format:
  html:
    embed-resources: true
---

Create a BAM file with manually written reads, so we know the expected result for testing.

Let's start with some initializations:

```{r}
library(tibble)

sq <- function(n){
  sample(c("A","C","T","G"), size = n, replace = TRUE) |>
    paste0(collapse = "")
}
qu <- function(n){
  sample(LETTERS, size = n, replace = TRUE) |>
    paste0(collapse = "")
}

write_sam <- function(reads, filename = "test.sam"){
  max_LN <- max(reads$pos + reads$tlen)*2
  header <- "@HD\tVN:1.4\tSO:coordinate"
  for(chrom in unique(reads$rname)){
    header <- c(header, paste0("@SQ\tSN:",chrom,"\tLN:",max_LN))
  }
  
  writeLines(header, filename, sep = "\n")
  readr::write_delim(reads, delim = '\t',
              col_names = FALSE, append = TRUE,
              file = filename, eol = "\n")
}

```


## nduo-6: + strand, single exon


```{text}
with # = 9.7 bp      5'                                               3'
                        #############################################
                        |                                           |
genomic               113                                         549
BED ref               112                                         549

   |bins spliced        0       100       200       300       400   |
5' |bins unspliced      0       100       200       300       400   |
   |bins genomic      113       213       313       413       513   |

   |bins spliced        |   400       300       200       100       0
3' |bins unspliced      |   400       300       200       100       0
   |bins genomic        |   149       249       349       449     549
                        #############################################
                        |     |   ^     |   ^     |   ^     |   ^   |
    genomic           113                                         549
    reads                  >>>>>>>>>>                    
                               >>>>...........>>>>      
                        #############################################
                        |     |   ^     |   ^     |   ^     |   ^   |
```

Note, looking at actual 10x data, the flags are `0`, `4`, `16`, `1024`, and `1040`.

 tag  |  Interpretation
------|----------------
   0  | -
   4  | unmapped
  16  | reverse strand
 1024 | duplicate
 1040 | duplicate on reverse strand

Note also that they all have `TLEN` of `0`, but here we keep the actual length. But we use the same `RNEXT` and `PNEXT`.

```{r}
reads1 <- tribble(
  ~qname, ~flag, ~rname, ~pos, ~mapq,  ~cigar, ~rnext, ~pnext, ~tlen,    ~seq, ~qual,
  "mt_1",    0, "MtDNA", 123,   255,  "100M",    "*",      0,   100, sq(100),qu(100),
  "mt_2",    0, "MtDNA", 160,   255,"40M130N70M","*",      0,   110, sq(110),qu(110)
  )
```


Hence the expected result:
```{text}
                      5'                                               3'
                        #############################################
                        |     |   ^     |   ^     |   ^     |   ^   |
    genomic           113                                         549
    reads                  >>>>>>>>>>                   
                               >>>>...........>>>>>>   
    5' counts           |    2    |    1    |    1    |    0    |   
    3' counts                 |    2    |    1    |    1    |   0   |
                        #############################################
                        |     |   ^     |   ^     |   ^     |   ^   |
                      113   149 213   249 313   349 413   449 513 549
                          
```




## B0348.5b.1: + strand, 2 exons

```{text}
with # = 20 bp      5'                                                          3'
                        #####################   ##############################
                        |                   |   |                            |
    genomic         5,536               5,966   6,024                    6,634
    BED ref         5,535                                                6,634
                        #####################   ##############################
                        |   |    |    |    |     |    |    |    |    |    |  
   |bins spliced        0 100  200  300  400   500  600  700  800  900 1000
5' |bins unspliced      0 100  200  300  400   557  657  757  857  957 1057
   |bins genomic     5536   . 5736 5836 5936  6093 6193 6293 6393 6493 6593
   
                        #####################   ##############################
                         |    |    |    |       |    |    |    |    |    |   |
   |bins spliced      1000  900  800  700     600  500  400  300  200  100   0
3' |bins unspliced    1057  957  857  757     600  500  400  300  200  100   0
   |bins genomic      5577 5677 5777 5877    6034    . 6234    . 6434    .6634
                        #####################   ##############################
                        l|  ^ |  ^ |  ^ |  ^l   |^   |^   |^   |^   |^   |^  l
    genomic         5,536               5,966   6,024                    6,634
    reads                    >>>>                           
                                   >>>>>>>>>>...>>>>>       
                        l   ^ |  ^ |  ^ |  ^l   |^   |^   |^   |^   |^   |^  l
    5' counts           | 0 | 1  |  1 |  1 |  1  |  1 |  0 | 0  |  0 | 0  |
    3' counts            |  1 |  2 |  1 |   1   |  1 |  0 | 0  |  0 |  0 | 0 |
```


```{r}
reads2 <- tribble(
  ~qname, ~flag, ~rname, ~pos, ~mapq,   ~cigar, ~rnext, ~pnext, ~tlen,  ~seq, ~qual,
  "b0_1",     0,    "V", 5640,   255,   "80M",     "*",      0,   80, sq(80), qu(80),
  "b0_2",     0,    "V", 5766,  255,"200M58N100M", "*",      0,  300,sq(300),qu(300)
)
```




## Y74C9A.6: - strand, single exon (note: snoRNA)


```{text}
with # = 3.3bp       3'                                                    5'
                        #################################################
                        |                                               |
    genomic         3,747                                           3,909
    BED ref         3,746                                           3,909
                        #################################################
                                        |                               |
   |bins spliced                      100                               0
5' |bins unspliced                    100                               0
   |bins genomic                    3,809                           3,909
                        #################################################
                                                    |
   |bins spliced        0                         100
3' |bins unspliced      0                         100
   |bins genomic     3747                        3847
                        l               ^           |                   l
    reads                  >>>>>>>>>>      >>>>>>>>
                                >>>>>>>>>>>>>>
    5' counts                           |                2               |
    3' counts           |             3             |
   
```


```{r}
reads3 <- tribble(~qname, ~flag, ~rname, ~pos, ~mapq, ~cigar, ~rnext, ~pnext, ~tlen,  ~seq, ~qual,
                  "y74_1",   16,     "I", 3751,   255, "33M",    "*",   0,   33, sq(33), qu(33),
                  "y74_2",   16,     "I", 3798,   255, "46M",     "*",  0,   46, sq(46), qu(46),
                  "y74_3",   16,     "I", 3810,   255, "26M",     "*",  0,   26, sq(26), qu(26))
```





## F23F1.10.1: - strand, two exons


```{text}
with # = 10bp       3'                                                      5'
                        #####################                ##############
                        |                   |                |            |
    genomic        41,471              41,682            41844       41,982
    BED ref        41,470                                            41,982
                        #####################                ##############
                            |         |                        |         |
   |bins spliced          300       200                      100         0
5' |bins unspliced        461       361                      100         0
   |bins genomic       41,521    41,621                   41,882    41,982
                        #####################                ##############
                                  |         |                         |     
   |bins spliced        0       100       200                       300
3' |bins unspliced      0       100       200                       461
   |bins genomic   41,471    41,571    41,671                    41,932
   
   reads                                 <<<<................<<<<<
                                        <<<
                        #####################                ##############
                        |   ^     |   ^     |                  ^      |   ^
   5' counts                |    0    |         2              |    1     |
   3' counts            |    0    |    2    |          1              |
```





```{r}
reads4 <- tribble(
  ~qname, ~flag, ~rname,  ~pos, ~mapq,       ~cigar, ~rnext, ~pnext, ~tlen,  ~seq, ~qual,
  "f23_1",   16,   "II", 41661,   255, "21M161N50M",    "*",      0,   71, sq(71), qu(71),
  "f23_2",   16,   "II", 41632,   255,        "30M",    "*",      0,   30, sq(30), qu(30)
)
```




## End

Now save this test file as a BAM.


```{r}
#| eval: false
write_sam(rbind(reads1, reads2, reads3, reads4),
          filename = "../tests/testthat/fixtures/simple.sam")

Rsamtools::asBam("../tests/testthat/fixtures/simple.sam", overwrite = TRUE)

```





# Different bins

We also want to check that the results stay valid with different sets of bins, using the same transcripts and reads as above.

Let's use this set:

```{r}
#| eval: false
bins <- c(0, 10, 20,60, 150, 300, 3000)
```


## F23F1.10.1: - strand, two exons


```{text}
with # = 10bp       3'                                                      5'
                        #####################                ##############
                        |                   |                |            |
    genomic        41,471              41,682            41844       41,982
    BED ref        41,470                                            41,982
                        #####################                ##############
                            |              |                        |   |||
   |bins spliced          300            150                       60  20 0
5' |bins unspliced        461            311                              0
   |bins genomic       41,521         41,671                   41,922     41,982
                        #####################                ##############
                        |||   |        |                              |     
   |bins spliced        0    60      150                            300
3' |bins unspliced      0    60      150                            461
   |bins genomic   41,471  41531  41,621                         41,932
   
   reads                                 <<<<................<<<<<
                                        <<<
                        #####################                ##############
                        |   ^ |            ^                        ^ |  ^^
   5' counts                |      2       |    1                   | 0  ||
                                                                         00
   3' counts            |||  0|     0  |          2                   |
                        00
```




