<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>test_descriptions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="test_descriptions_files/libs/clipboard/clipboard.min.js"></script>
<script src="test_descriptions_files/libs/quarto-html/quarto.js"></script>
<script src="test_descriptions_files/libs/quarto-html/popper.min.js"></script>
<script src="test_descriptions_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="test_descriptions_files/libs/quarto-html/anchor.min.js"></script>
<link href="test_descriptions_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="test_descriptions_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="test_descriptions_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="test_descriptions_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="test_descriptions_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">



<section id="examples-and-tests" class="level1">
<h1>Examples and tests</h1>
<p>This file contains some illustrated examples that help think about the low-level functions.</p>
<p>The reference file was generated with the code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>tx_to_keep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"MTCE.3.1"</span>, <span class="st">"B0348.5b.1"</span>, <span class="st">"C06C3.1b.1"</span>, <span class="st">"Y74C9A.6"</span>, <span class="st">"F23F1.10.1"</span>, <span class="st">"T14G10.2b.1"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>raw <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">"data/geneBody_coverage/c_elegans.PRJNA13758.WS281.canonical_geneset.genePred.bed"</span>) <span class="sc">|&gt;</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_match</span>(<span class="st">"^[IVXMtDNA]{1,5}</span><span class="sc">\t\\</span><span class="st">d{1,9}</span><span class="sc">\t\\</span><span class="st">d{1,9}</span><span class="sc">\t</span><span class="st">([A-Za-z0-9</span><span class="sc">\\</span><span class="st">.]+)</span><span class="sc">\t</span><span class="st">.*"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">writeLines</span>(raw[raw[,<span class="dv">2</span>] <span class="sc">%in%</span> tx_to_keep, <span class="dv">1</span>],</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>           <span class="st">"data/geneBody_coverage/c_elegans.PRJNA13758.WS281_filtered.bed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../R/bin_coords_conversions.R"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(GenomicRanges))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>chrom_list <span class="ot">&lt;-</span> <span class="fu">as.roman</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">as.character</span>() <span class="sc">|&gt;</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">c</span>(<span class="st">"X"</span>,<span class="st">"MtDNA"</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>opt <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>,<span class="dv">400</span>,<span class="dv">500</span>,<span class="dv">600</span>,<span class="dv">700</span>,<span class="dv">800</span>,<span class="dv">900</span>,<span class="dv">1000</span>,<span class="dv">1100</span>,<span class="dv">1200</span>,<span class="dv">1300</span>,<span class="dv">1400</span>,<span class="dv">1500</span>,<span class="dv">1600</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>          <span class="dv">1700</span>,<span class="dv">1800</span>,<span class="dv">1900</span>,<span class="dv">2000</span>,<span class="dv">2100</span>,<span class="dv">2200</span>,<span class="dv">2300</span>,<span class="dv">2400</span>,<span class="dv">2500</span>,<span class="dv">2700</span>,<span class="dv">2900</span>,<span class="dv">3100</span>,<span class="dv">3400</span>,<span class="dv">3700</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>          <span class="dv">4000</span>,<span class="dv">4500</span>,<span class="dv">5000</span>,<span class="dv">5500</span>,<span class="dv">6000</span>,<span class="dv">6500</span>,<span class="dv">7000</span>,<span class="dv">8000</span>,<span class="dv">9000</span>,<span class="dv">10000</span>,<span class="cn">Inf</span>)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>gene_struct <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"../tests/testthat/fixtures/c_elegans.PRJNA13758.WS281_filtered.bed"</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>                          <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>                        <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">"chr"</span>,<span class="st">"transcript_start"</span>,<span class="st">"transcript_end"</span>,<span class="st">"transcript_name"</span>,<span class="st">"score"</span>,</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"strand"</span>,<span class="st">"thick_start"</span>,<span class="st">"thick_end"</span>,<span class="st">"rgb"</span>,</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"nb_exons"</span>,<span class="st">"exons_lengths"</span>,<span class="st">"exons_starts"</span>),</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>                        <span class="at">col_types =</span> <span class="fu">cols</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>                          <span class="at">chr =</span> <span class="fu">col_factor</span>(chrom_list),        <span class="co">#1</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                          <span class="at">transcript_start =</span> <span class="fu">col_integer</span>(),        <span class="co">#2 on the + strand (even if tx on -)</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                          <span class="at">transcript_end =</span> <span class="fu">col_integer</span>(),          <span class="co">#3</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                          <span class="at">transcript_name =</span> <span class="fu">col_character</span>(),       <span class="co">#4</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>                          <span class="at">score =</span> <span class="fu">col_factor</span>(<span class="st">"0"</span>),                 <span class="co">#5 ignore</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>                          <span class="at">strand =</span> <span class="fu">col_factor</span>(<span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>)),         <span class="co">#6</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>                          <span class="at">thick_start =</span> <span class="fu">col_integer</span>(),             <span class="co">#7 ignore</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>                          <span class="at">thick_end =</span> <span class="fu">col_integer</span>(),               <span class="co">#8 ignore</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rgb =</span> <span class="fu">col_factor</span>(<span class="st">"0"</span>),                   <span class="co">#9 ignore</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>                          <span class="at">nb_exons =</span> <span class="fu">col_integer</span>(),                <span class="co">#10</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>                          <span class="at">exons_lengths =</span> <span class="fu">col_character</span>(),         <span class="co">#11</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>                          <span class="at">exons_starts =</span> <span class="fu">col_character</span>()           <span class="co">#12 on the + strand (even if tx on -)</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>                        )) <span class="sc">|&gt;</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">exons_lengths =</span> <span class="fu">strsplit</span>(exons_lengths, <span class="st">","</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>           <span class="fu">map</span>(as.integer),</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>         <span class="at">exons_starts =</span> <span class="fu">strsplit</span>(exons_starts, <span class="st">","</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>           <span class="fu">map</span>(as.integer))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="nduo-6-strand-single-exon" class="level2">
<h2 class="anchored" data-anchor-id="nduo-6-strand-single-exon">nduo-6: + strand, single exon</h2>
<pre><code>with # = 9.7 bp      5'                                               3'
                        #############################################
                        |                                           |
    genomic           113                                         549
    BED ref           112                                         549
  
   |bins spliced        0       100       200       300       400   |
5' |bins unspliced      0       100       200       300       400   |
   |bins genomic      113       213       313       413       513   |
  
   |bins spliced        |   400       300       200       100       0
3' |bins unspliced      |   400       300       200       100       0
   |bins genomic        |   149       249       349       449     549</code></pre>
<section id="from-5-end" class="level3">
<h3 class="anchored" data-anchor-id="from-5-end">From 5’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"MTCE.3.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 5L</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300 400</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300 400</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 113 213 313 413 513</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   114-213      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   214-313      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   314-413      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[4]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   414-513      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="from-3-end" class="level3">
<h3 class="anchored" data-anchor-id="from-3-end">From 3’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"MTCE.3.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300 400</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300 400</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 549 449 349 249 149</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   450-549      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   350-449      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   250-349      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[4]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]    MtDNA   150-249      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
<section id="b0348.5b.1-strand-2-exons" class="level2">
<h2 class="anchored" data-anchor-id="b0348.5b.1-strand-2-exons">B0348.5b.1: + strand, 2 exons</h2>
<pre><code>with # = 20 bp      5'                                                          3'
                        #####################   ##############################
                        |                   |   |                            |
    genomic         5,536               5,966   6,024                    6,634
    BED ref         5,535                                                6,634
                        #####################   ##############################
                        |   |    |    |    |     |    |    |    |    |    |  
   |bins spliced        0 100  200  300  400   500  600  700  800  900 1000
5' |bins unspliced      0 100  200  300  400   557  657  757  857  957 1057
   |bins genomic     5536   . 5736 5836 5936  6093 6193 6293 6393 6493 6593
   
                        #####################   ##############################
                         |    |    |    |       |    |    |    |    |    |   |
   |bins spliced      1000  900  800  700     600  500  400  300  200  100   0
3' |bins unspliced    1057  957  857  757     600  500  400  300  200  100   0
   |bins genomic      5577 5677 5777 5877    6034    . 6234    . 6434    .6634</code></pre>
<section id="from-5-end-1" class="level3">
<h3 class="anchored" data-anchor-id="from-5-end-1">From 5’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"B0348.5b.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 5L</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500  600  700  800  900 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  557  657  757  857  957 1057</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5536 5636 5736 5836 5936 6093 6193 6293 6393 6493 6593</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5537-5636      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5637-5736      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5737-5836      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[4]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5837-5936      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[5]]
GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5937-5966      +
  [2]        V 6024-6093      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[6]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6094-6193      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[7]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6194-6293      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[8]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6294-6393      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[9]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6394-6493      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[10]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6494-6593      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="from-3-end-1" class="level3">
<h3 class="anchored" data-anchor-id="from-3-end-1">From 3’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"B0348.5b.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500  600  700  800  900 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500  600  757  857  957 1057</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 6634 6534 6434 6334 6234 6134 6034 5877 5777 5677 5577</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6535-6634      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6435-6534      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6335-6434      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[4]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6235-6334      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[5]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6135-6234      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[6]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 6035-6134      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[7]]
GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5878-5966      +
  [2]        V 6024-6034      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[8]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5778-5877      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[9]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5678-5777      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[10]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        V 5578-5677      +
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
<section id="y74c9a.6---strand-single-exon-note-snorna" class="level2">
<h2 class="anchored" data-anchor-id="y74c9a.6---strand-single-exon-note-snorna">Y74C9A.6: - strand, single exon (note: snoRNA)</h2>
<pre><code>with # = 3.3bp       3'                                                    5'
                        #################################################
                        |                                               |
    genomic         3,747                                           3,909
    BED ref         3,746                                           3,909
                        #################################################
                                        |                               |
   |bins spliced                      100                               0
5' |bins unspliced                    100                               0
   |bins genomic                    3,809                           3,909
                        #################################################
                                                    |
   |bins spliced        0                         100
3' |bins unspliced      0                         100
   |bins genomic     3747                        3847</code></pre>
<section id="from-5-end-2" class="level3">
<h3 class="anchored" data-anchor-id="from-5-end-2">From 5’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"Y74C9A.6"</span>) <span class="sc">|&gt;</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 5L</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3909 3809</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        I 3810-3909      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="from-3-end-2" class="level3">
<h3 class="anchored" data-anchor-id="from-3-end-2">From 3’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"Y74C9A.6"</span>) <span class="sc">|&gt;</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3747 3847</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]        I 3748-3847      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
<section id="f23f1.10.1---strand-two-exons" class="level2">
<h2 class="anchored" data-anchor-id="f23f1.10.1---strand-two-exons">F23F1.10.1: - strand, two exons</h2>
<pre><code>with # = 10bp       3'                                                      5'
                        #####################                ##############
                        |                                                 |
    genomic        41,471                                            41,982
    BED ref        41,470                                            41,982
                        #####################                ##############
                             |         |                         |       |
   |bins spliced           300       200                       100       0
5' |bins unspliced         461       361                       100       0
   |bins genomic        41,521    41,621                    41,882  41,982
                        #####################                ##############
                                 |         |                        |     
   |bins spliced        0      100       200                      300
3' |bins unspliced      0      100       200                      461
   |bins genomic   41,471   41,571    41,671                   41,932</code></pre>
<section id="from-5-end-3" class="level3">
<h3 class="anchored" data-anchor-id="from-5-end-3">From 5’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"F23F1.10.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 5L</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 361 461</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 41982 41882 41621 41521</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41883-41982      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 2 ranges and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41622-41682      -
  [2]       II 41844-41882      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 1 range and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41522-41621      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="from-3-end-3" class="level3">
<h3 class="anchored" data-anchor-id="from-3-end-3">From 3’ end</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"F23F1.10.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 300</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 100 200 461</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 41471 41571 41671 41932</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
GRanges object with 1 range and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41472-41571      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[2]]
GRanges object with 1 range and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41572-41671      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths

[[3]]
GRanges object with 2 ranges and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]       II 41672-41682      -
  [2]       II 41844-41932      -
  -------
  seqinfo: 7 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
<section id="mel-11-strand-many-exons" class="level2">
<h2 class="anchored" data-anchor-id="mel-11-strand-many-exons">mel-11: + strand, many exons</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"C06C3.1b.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300 1400
[16] 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2700 2900 3100 3400</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500 1142 1242 1391 1534 2929 3029 3322 3422 4081
[16] 4181 5170 5270 5754 5854 6012 6112 6806 6906 7006 7283 7887 8087 8287 9151</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 9369099 9368999 9368899 9368799 9368699 9368599 9367957 9367857 9367708
[10] 9367565 9366170 9366070 9365777 9365677 9365018 9364918 9363929 9363829
[19] 9363345 9363245 9363087 9362987 9362293 9362193 9362093 9361816 9361212
[28] 9361012 9360812 9359948</code></pre>
</div>
</div>
</section>
<section id="pxf-1---strand-many-exons" class="level2">
<h2 class="anchored" data-anchor-id="pxf-1---strand-many-exons">pxf-1: - strand, many exons</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> gene_struct <span class="sc">|&gt;</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transcript_name <span class="sc">==</span> <span class="st">"T14G10.2b.1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tx_width =</span> <span class="fu">map_int</span>(exons_lengths, sum)) <span class="sc">|&gt;</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.list</span>()</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>opt<span class="sc">$</span>endedness <span class="ot">&lt;-</span> 3L</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords <span class="ot">&lt;-</span> <span class="fu">map</span>(xx<span class="sc">$</span>tx_width, <span class="sc">~</span>bins[bins <span class="sc">&lt;=</span> .x])[[<span class="dv">1</span>]]</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_unspliced</span>(bins_in_spliced_coords,</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>strand,</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]],</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>                                                xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]])</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords <span class="ot">&lt;-</span> <span class="fu">coords_to_genomic</span>(xx<span class="sc">$</span>transcript_start,</span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>transcript_end,</span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a>  bins_in_unspliced_coords,</span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>strand)</span>
<span id="cb77-20"><a href="#cb77-20" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">coords_as_granges</span>(xx<span class="sc">$</span>chr, bins_in_genomic_coords, xx<span class="sc">$</span>strand, xx<span class="sc">$</span>transcript_start, xx<span class="sc">$</span>exons_starts[[<span class="dv">1</span>]], xx<span class="sc">$</span>exons_lengths[[<span class="dv">1</span>]])</span>
<span id="cb77-21"><a href="#cb77-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-22"><a href="#cb77-22" aria-hidden="true" tabindex="-1"></a>bins_in_spliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300 1400
[16] 1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2700 2900 3100 3400
[31] 3700 4000 4500</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>bins_in_unspliced_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]    0  100  200  300  449  549  649 1134 1234 1470 1682 1782 2229 2329 2744
[16] 2844 2944 3094 3194 3341 3441 3541 3686 3786 3942 4042 4242 4442 4690 4990
[31] 5290 5590 6521</code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>bins_in_genomic_coords</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 10164288 10164388 10164488 10164588 10164737 10164837 10164937 10165422
 [9] 10165522 10165758 10165970 10166070 10166517 10166617 10167032 10167132
[17] 10167232 10167382 10167482 10167629 10167729 10167829 10167974 10168074
[25] 10168230 10168330 10168530 10168730 10168978 10169278 10169578 10169878
[33] 10170809</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>